Implementation of KS legacy with custom signals

rsheer opened the pull request:
This is implementation to the proposed interface change for the kitchen-sink legacy component.

Following changes had to be done to make this happen:
- Implementation of the 2 interface methods (of course)
- Add an argument of the type `SignalGeneratorInterface` to the constructor
  - this had to be done, so Signals with unique IDs could be created
- Due to the constructor changes, following classes needed some adjustments to create an instance of legacy component implementation:
  - `UI\Implementation\Factory` -> Factory which creates an instance of `LegacyImplementation`
  - `GlobalScreen\Scope\Layout\Provider\PagePart\StandardPagePartProvider` -> somehow creates an instance of `LegacyImplementation` as well
- Changes in the renderer, so the signals will also be rendered with the legacy component
  - I had to implement a method called `getAllSignals` to the legacy implementation (but not the interface), so I was able to get all registered signals for the rendering.
- Added some unit tests for the new implemented methods

I'm open for all the questions and feedback  :)

rsheer committed:
KS Legacy custom signal: Implemented new interface functionality

rsheer committed:
KS Legacy custom signal: Changed signature of legacy impl constructor

rsheer committed:
KS Legacy custom signal: Add method to interface to get all signals

rsheer committed:
KS Legacy custom signal: Implemented rendering for custom signals

rsheer committed:
KS Legacy custom signal: Added unit tests for custom signals

rsheer committed:
Revert "KS Legacy custom signal: Add method to interface to get all signals"

This reverts commit 7409b460c80e3edfa1e8cd559c1f3f1b00be7e32.

The method to get all custom signals should only be in the implementation, not in the interface

rsheer committed:
KS Legacy custom signal: Added phpdoc comments to some implemented methods

rsheer commented:
Question: As seen with the failed tests, the change of the constructors signature was more impactful than expected. What do you suggest me? Should I get rid of the dependency injection and just create a SignalGenerator in the constructor or does anyone have another suggestions?

nhaagen commented on src/GlobalScreen/Scope/Layout/Provider/PagePart/StandardPagePartProvider.php:
```diff
@@ -55,7 +56,7 @@ public function __construct()
     public function getContent() : ?Legacy
     {
-        return $this->content ?? new LegacyImplementation("");
+        return $this->content ?? new LegacyImplementation("", new SignalGenerator());
```
why not use $this->ui->legacy() instead of "new Legacy" ?

nhaagen commented on src/UI/Implementation/Component/Legacy/Renderer.php:
```diff
@@ -34,4 +35,29 @@ protected function getComponentInterfaceName()
+        {
+            $code .= "$(this).on('{$signal['signal']->getId()}', function(e){".$signal['js_code']."});";
+        }
+        $code .= "</script>";
```
Please bind with "withAdditionalOnLoadCode".

nhaagen commented on src/UI/Implementation/Component/Legacy/Renderer.php:
```diff
@@ -24,7 +25,7 @@ public function render(Component\Component $component, RendererInterface $defaul
         $this->checkComponent($component);
 
-        return $component->getContent();
+        return $component->getContent() . $this->renderCustomSignals($component);
```
Please use  "withAdditionalOnLoadCode" to inject JS-code.

nhaagen reviewed (changes requested):
Hi @rsheer ,
thanks for the implementation. Your code looks well structured!

As to your question: Impacts will radically disappear by using the factory to construct a Component - this was/is bad in the StandardPagePartProvider already and you only happened to break it ;)

You actually turned Legacy into a JavaScriptBindable - there is both an interface as well as a trait. JavaScriptBindable provides withAdditionalOnLoadCode - please use this to inject JS-code.
Maybe have a look at the button's renderer? 

Best regards,
Nils

klees reviewed (changes requested):
Hi @rsheer!

Thanks again for taking care of this so dilligently.

Please implement the following changes:

* [x] Use `$this->ui->legacy()` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2274/files#diff-bf0ada09944cf7a62b91f605659664f3R59) as @nhaagen mentions to reduce the coupling between the Global Screen and the UI Framework to the public interface of the UI Framework.
* [x] Use `withAdditionalOnLoadCode` to bind js-code to DOM-nodes as @nhaagen mentions. `withAdditionalOnLoadCode` is the only low level portal to the client side that the UI Framework supports. You could, e.g., have a look into the `Renderer` of the Button for a usage example.
* [x] Do not create a new `SignalGenerator` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2274/files#diff-f9c5461bddd130c3d60deb1836292034R233) but instead inject it via the factory, like (e.g.) the factory of `ViewControl` does.

Ping me if you are ready for a final review.

Best regards!

rsheer committed:
KS Legacy custom signal: Use factory to create legacy component

rsheer committed:
KS Legacy custom signal: Legacy component extends now from JavaScriptBindable Interface

rsheer committed:
KS Legacy custom signal: Created Legacy Facory, so the signal generator can be injected in it

rsheer committed:
KS Legacy custom signal: Implemented rendering for js code

rsheer commented:
Hey @nhaagen & @klees!

Thank you very much for your feedback and the review! The requested changes are now implemented and pushed to this PR. But it seems that I have some other things that broke as well...

To inject the SignalGenerator, I had to create a Factory for the Legacy Component. All other components which are using the `SignalGenerator` as well are having a factory since there are multiple variations of them (e.g. button -> standard, primary etc.). But since there is just one variation for the Legacy component and `$ui_factory()->legacy_factory()->legacy("")` looks awful to me, the Legacy Factory is used inside of the [UI Factory](https://github.com/ILIAS-eLearning/ILIAS/pull/2274/files#diff-f9c5461bddd130c3d60deb1836292034R240). I hope this move is ok to you.

nhaagen committed:
UI Legacy via factory, adjusting tests

nhaagen commented:
Hey @rsheer ,
you sure went the extra mile ;) 
Although all feasible and correct, the non-optional 2nd parameter (SignalGenerator) 
in the constructor of Legacy still breaks a lot of older parts, especially in the tests.

Please bear with me in that I'm not all decided on possible solutions yet and will still 
need some time to look into this more deeply.

Best regards so far, I will get back to you early next week.

rsheer commented:
Hey @nhaagen,

thanks for your feedback. I appreciate that!

I have looked into the Unit Tests and could make out 3 scenarios, where and why they fail:
- The unit tests are creating an instance of the legacy component, which needs a signal generator in the constructor
  - Possible Solution -> Use a mocked `SignalGeneratorInterface` for the component constructor

- The unit tests are creating a UI factory, which needs a legacy factory in the constructor
  - Possible Solution -> Use a legacy factory with a mocked `SignalGeneratorInterface` in the constructor

- The unit tests are comparing their output against a hard coded html string. But since the legacy renderer calls `$this->bindJavaScript($legacy)`, the IDs for the html tags (e.g. `<div id="id_2">` instead of `id_1`) are getting mixed up. Even without creating a custom signal.
  - Possible Solution 1 -> Change the hard coded IDs for every validation string
  - Possible Solution 2 -> Only use `bindJavaScript`  in the legacy renderer, if the legacy component has at least one signal. Looks something like this: `if(count($component->getAllCustomSignals()) > 0) $this->bindJavaScript($component)` 

Well, I'm not an expert in the topic of unit testing. It seems like a lot of work to just change all constructors. Especially for the fact that this methods are already deprecated during development and the changes have to be reverted in a year or two... But I think those changes are a way better solution than just making the signal generator an optional argument in the constructor.

Best regards

klees commented:
Hi @rsheer,

the unit tests of the UI-framework are really bad when it comes to looking into the actually rendered html. Unfortunately we do not have a better solution there then comparing strings, but this is extremely brittle and thus a bad unit test.

Sry...

Amstutz commented:
Conserning the signal id issue in unit tests, I usually do the following pattern to sail around this.

Provide a UIFactory, providing you with all components needed for your tests and extending the NoUIFactory, to spare myself from some further footwork. Then injecting the respective components with a mocked sig_generator:

```
    public function getUIFactory()
    {
        $factory = new class extends NoUIFactory {
            //E.g. if symbols are needed
            public function symbol() : ILIAS\UI\Component\Symbol\Factory
            {
                return new I\Symbol\Factory(
                    new I\Symbol\Icon\Factory(),
                    new I\Symbol\Glyph\Factory()
                );
            }
           //Note that we inject this here.
            public function mainControls() : C\MainControls\Factory
            {
                return new I\MainControls\Factory($this->sig_gen,
                    new I\MainControls\Slate\Factory($this->sig_gen
                        , new \ILIAS\UI\Implementation\Component\Counter\Factory())
                );
            }
        };

        //Kick the mocked generator into the scope of your mocked factory.
        $factory->sig_gen = Mockery::mock(I\SignalGeneratorInterface::class);
        $factory->sig_gen->shouldReceive("create")->andReturn(new I\Signal("id"));
        //There you go
        return $factory;
    }
```

@klees Should we promote this a bit further up the chain and maybe even extend the NoUIFactory a bit to offer this work by default?

nhaagen commented:
+1 for extending NoUIFactory; 
There are still quite a few places where the actual UI-Factory is constructed with a lot of mocks. 
These should go, imo.

rsheer commented:
Hey everyone,

thanks for your feedback! The unit test with the NoUIFactory looks good to me.

So what should I / we do now next? Should I change the breaking unit tests to the NoUIFactory?

Thx for your help and feedback. I've already learnt a lot about the UI framework and I really like to work with it!

nhaagen committed:
UI Legacy, extend legacy-construction with signal generator, fixing some tests

rsheer committed:
Merge pull request #1 from nhaagen/pr_legacysignals

UI, Legacy with Signals, some tests

rsheer committed:
UI: Fixed Unit Tests in RepositoryObjectTest

rsheer committed:
KS Legacy: Fixed hardcoded id=id_X in unit test

This had to be done, since the fact, that legacy now implements JavaScriptBindable has an influence on the generated IDs in the rendered HTML code

rsheer committed:
KS Legacy: Fixed some more hardcoded id=id_X in unit test

rsheer commented:
The final unit tests are repaired and everything seems to work now. Therefore, I'm ready for the final review. :)

klees commented:
Thanks @rsheer. Please note, that I squashed your commits into one.
